### PR TITLE
Convert System.Security.Cryptography.Native to C

### PR DESCRIPTION
**Pull request by kasper3**

### CODE REVIEW

**Comment by kasper3** [comment]
test NETFX x86 Release Build please

Unrelated transient error, that was earlier reported at https://github.com/dotnet/corefx/issues/28816

https://mc.dot.net/#/user/kasper3/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/f32e137e22f3c58e581c30e7f56e8892584fbf41/workItem/System.Runtime.Serialization.Formatters.Tests

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c [hunk 1]
```diff
@@ -1332,8 +1331,8 @@ extern "C" int32_t CryptoNative_EnsureOpenSslInitialized()
     }
 
     // Create the locks array
-    g_locks = new (std::nothrow) pthread_mutex_t[numLocks];
-    if (g_locks == nullptr)
+    g_locks = (pthread_mutex_t*)malloc(sizeof(pthread_mutex_t) * (size_t)numLocks);
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Is there any situation where we'd need to worry about overflows here?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_evp_cipher.c [hunk 2]
```diff
@@ -5,76 +5,75 @@
...
-    EVP_CIPHER_CTX_init(ctx.get());
+    EVP_CIPHER_CTX_init(ctx);
 
     // Perform partial initialization so we can set the key lengths
-    int ret = EVP_CipherInit_ex(ctx.get(), type, nullptr, nullptr, nullptr, 0);
+    int ret = EVP_CipherInit_ex(ctx, type, NULL, NULL, NULL, 0);
     if (!ret)
     {
-        return nullptr;
+        return NULL;
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_evp_cipher.c]: Don't we need to free the ctx in all of these cases where it's been allocated but we return NULL?  Previously it looks like the unique_ptr was handling that.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c [hunk 3]
```diff
@@ -7,82 +7,81 @@
...
 
-    HMAC_CTX_init(ctx.get());
-    int ret = HMAC_Init_ex(ctx.get(), key, keyLen, md, nullptr);
+    HMAC_CTX_init(ctx);
+    int ret = HMAC_Init_ex(ctx, key, keyLen, md, NULL);
 
     if (!ret)
     {
-        return nullptr;
+        return NULL;
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c]: Same unique_ptr question.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 4]
```diff
@@ -326,55 +327,56 @@ static bool GetDescriptionValue(
...
-                                                CipherAlgorithmType& dataCipherAlg,
-                                                ExchangeAlgorithmType& keyExchangeAlg,
-                                                HashAlgorithmType& dataHashAlg,
-                                                DataHashSize& hashKeySize)
+                                                CipherAlgorithmType* dataCipherAlg,
+                                                ExchangeAlgorithmType* keyExchangeAlg,
+                                                HashAlgorithmType* dataHashAlg,
+                                                DataHashSize* hashKeySize)
 {
-    const int descriptionLength = 256;
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: Why did you make this a #define?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c [hunk 5]
```diff
@@ -326,55 +327,56 @@ static bool GetDescriptionValue(
...
-    char description[descriptionLength] = {};
+    char description[descriptionLength];
     SSL_CIPHER_description(cipher, description, descriptionLength - 1); // ensure description is NULL-terminated
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: Previously the whole array was being initialized to zero, which is how it was ensuring that it was NULL-terminated.  After your change, that's no longer the case.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c (hunk 4 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: This is used in statically-bounded array declaration few lines below. In C, `const` is a type as opposed to C++ and array declaration requires a constant expression, so compiler was complaining about something related to runtime-bounded array or variable-length array from ISO C99.
This is the one of few ways of definition the contant in C:

1. `extern const int MYCONST = 256` 
2. `#define MYCONST 256`
3. `enum { MYCONST = 256}`
4.. `256`
5. `sizeof(some_construct_that_is_256_in_size)`

File: src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    uint8_t _;
+    if (keyLen == 0)
+        key = &_;
+
+    HMAC_CTX_init(ctx);
+    int ret = HMAC_Init_ex(ctx, key, keyLen, md, NULL);
+
+    if (!ret)
+    {
+        return NULL;
```
Comment (bartonjs) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c]: Another source where a free was lost when shifting off of unique_ptr.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h [hunk 7]
```diff
@@ -87,19 +88,19 @@ enum class HashAlgorithmType : int32_t
...
     SHA1_HashKeySize = 8 * SHA_DIGEST_LENGTH,
     SHA256_HashKeySize = 8 * SHA256_DIGEST_LENGTH,
     SHA384_HashKeySize = 8 * SHA384_DIGEST_LENGTH,
     GOST_HashKeySize = 256,
     Default = 0,
-};
+} DataHashSize;
 
-enum SslErrorCode : int32_t
+enum SslErrorCode
```
Comment (bartonjs) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h]: I think every other enum became `typedef enum { ... } Name;`, why is this one `enum Name { ... }`?

Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h]: Where C++ enum name was used in function signature, for its C counterpart, we needed to either do `typedef enum <optionalPrototypeName> {} Name` or declare from prototype / blueprint `enum <optionalPrototypeName> {} Name`.

In this PR and previous ones, I have only updated those `enum`  and `struct` to typedef which were used in some function signature and left rest with blueprint name only.
We can sweep all these `enum` and `struct` to one style regardless of where they are used. I can do it here or in the very last PR of this series where all the __cplusplus dedcode and C++ compiler flags from CMake will be cleaned up. We are one step away from that (pending only `System.Security.Cryptography.Native.Apple` under Unix after this PR).

Comment (bartonjs) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h]: Good enough for me :)

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c (hunk 5 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c]: Great, I have fixed it by adding `= {0}`.

In C++, empty brace `{}` is allowed but not in C as noted in http://en.cppreference.com/w/c/language/array_initialization. In C and C++ both, `= {0}` is a valid syntax. However, in C's global scope (outside of function body), declaration without initialization zeros out the array: https://stackoverflow.com/a/2589751.

**Comment by kasper3** [comment]
test NETFX x86 Release Build please

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c [hunk 8]
```diff
@@ -336,7 +335,7 @@ extern "C" int32_t CryptoNative_GetAsn1StringBytes(ASN1_STRING* asn1, uint8_t* p
         return -length;
     }
 
-    memcpy_s(pBuf, UnsignedCast(cBuf), asn1->data, UnsignedCast(length));
+    memcpy_s(pBuf, (size_t)cBuf, asn1->data, (size_t)length);
```
Comment (janvorli) [inline src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: The UnsignedCast contains an assert that the value we are casting is >= 0. With this change, we are losing that. It seems we should rather define a C version of the UnsignedCast to keep that debugging behavior.

**Comment by kasper3** [comment]
Note that `git mv` has a limit, if you move the file then change it afterwards, and that delta change is greater than some 3K chars (i don't remember exactly there is a hardcoded limit), then it shows as if file was removed and added. It doesn't happen with every file, only when subsequent change exceeds that hardcoded threshold.

We have covered that here https://github.com/dotnet/corefx/pull/30229#issue-193769969 (first two comments). So we have two commits, one is moving the files and one is changing the files. Please review commit-wise.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_bio.c [hunk 9]
```diff
@@ -6,49 +6,49 @@
 
 #include <assert.h>
 
-extern "C" BIO* CryptoNative_CreateMemoryBio()
+BIO* CryptoNative_CreateMemoryBio()
```
Comment (janvorli) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_bio.c]: Add `void` inside the parentheses for consistency with other places where you have added the `void`?

Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_bio.c]: Compiler only requires it in first declaration and subsequent forward declaration. It hasn't been done in previous PRs, and not a compile time error, should we skip it for now?

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c (hunk 1 above)
Comment (stephentoub) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: That's not what I was talking about. Multiplication can overflow back to being a positive number, at which point the malloc would succeed but for less space than was actually needed, which can then result in buffer overruns. That's very unlikely to happen here, but I'd still feel better if we handled it. We have a helper for doing safe multiplication.

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c (hunk 8 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Makes sense. Something that hasn't been taken care of in earlier PRs of same sorts. I can do it now or I can do it in the next to the next PR, that I mentioned to @bartonjs https://github.com/dotnet/corefx/pull/30376#discussion_r195300339. Whichever you prefer :)

File: src/Native/Unix/System.Security.Cryptography.Native/pal_bio.c (hunk 9 above)
Comment (janvorli) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_bio.c]: Ok, we can skip it.

**Comment by ViktorHofer** [comment]
@dotnet-bot test NETFX x86 Release Build

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c (hunk 8 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Do you have a recommendation how should the C version of UnsignedCast and SignedCast looks like?

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c (hunk 1 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Thanks for the explanation @stephentoub. I have pushed a commit using multiply_s.

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c (hunk 8 above)
Comment (janvorli) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Just assert that the input is in the right range and return the cast value. You'll obviously need to create couple of variants based on the types used in the casts.

Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: Thanks, i have pushed a commit.

**Comment by kasper3** [comment]
test NETFX x86 Release Build
test Packaging All Configurations x64 Debug Build

**Comment by ViktorHofer** [comment]
Please ignore failing System.Runtime.Serialization.Formatters tests in the netfx leg.

File: src/Native/Unix/Common/pal_utilities.h [hunk 10]
```diff
@@ -196,3 +197,21 @@ static inline bool CheckInterrupted(int32_t result)
 }
 
 #endif // __cplusplus
+
+inline static uint32_t UnsignedCast_Int32ToUint32(int32_t value)
```
Comment (janvorli) [inline src/Native/Unix/Common/pal_utilities.h]: @bartonjs, @stephentoub do you prefer this long function names or would you like shorter ones better (e.g. just Int32ToUInt32, etc.)?

Comment (bartonjs) [reply src/Native/Unix/Common/pal_utilities.h]: I'd go with `Int32ToUInt32`.  It's descriptive enough.

Comment (kasper3) [reply src/Native/Unix/Common/pal_utilities.h]: I've simplified the names.

File: src/Native/Unix/System.Security.Cryptography.Native/openssl.c [hunk 11]
```diff
@@ -386,7 +386,7 @@ int32_t CryptoNative_GetX509NameRawBytes(X509_NAME* x509Name, uint8_t* pBuf, int
         return -length;
     }
 
-    memcpy_s(pBuf, (size_t)cBuf, x509Name->bytes->data, (size_t)length);
+    memcpy_s(pBuf, Int32ToUint32(cBuf), x509Name->bytes->data, Int32ToUint32(length));
```
Comment (janvorli) [inline src/Native/Unix/System.Security.Cryptography.Native/openssl.c]: This should use the Int32ToSizeT

**Comment by kasper3** [comment]
@bartonjs, @janvorli, I have addressed all the comments. Can you take another look please?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h (hunk 7 above)
Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h]: I didn't realized that System.Native had some pending items, issued https://github.com/dotnet/corefx/pull/30431. Now `S.S.C.N.Apple` is the only one :)

Comment (kasper3) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h]: This is now addressed in https://github.com/dotnet/corefx/pull/30495. I have streamlined all `typedef struct` and `typedef enum`.

**Pull request merged**